### PR TITLE
Download epic

### PR DESCRIPTION
**Pull request by rgbkrk**
This moves the notebook file downloading into an epic, freeing the menu up from re-rendering on every single keystroke in the notebook...

### CODE REVIEW

**Comment by rgbkrk** [comment]
Gotta fix up the tests then this is good to go.

**Comment by rgbkrk** [comment]
Super good to go now!

File: packages/core/src/actions.js [hunk 1]
```diff
@@ -138,7 +57,7 @@ export const fetchContent = (payload: {
   params: Object,
   kernelRef: KernelRef,
   contentRef: ContentRef
-}): FetchContent => ({
+}): actionTypes.FetchContent => ({
```
Comment (theengineear) [inline packages/core/src/actions.js:60]: Oh dang! Good idea!

File: packages/core/__tests__/components/notebook-menu-extra-handlers-spec.js [hunk 2]
```diff
@@ -1,39 +0,0 @@
-// @flow
```
Comment (rgbkrk) [inline packages/core/__tests__/components/notebook-menu-extra-handlers-spec.js:1]: Note -- these tests got moved into the contents tests (of which there were none previously).

File: packages/core/src/components/notebook-menu/index.js [hunk 3]
```diff
@@ -407,23 +403,15 @@ const mapStateToProps = state => {
...
-    filepath: selectors.currentFilepath(state),
-    notebook: model.notebook,
     currentKernelRef: selectors.currentKernelRef(state),
     currentContentRef: contentRef
   };
 };
 
 const mapDispatchToProps = dispatch => ({
   saveNotebook: (payload: *) => dispatch(actions.save(payload)),
+  downloadNotebook: (payload: *) => dispatch(actions.downloadContent(payload)),
```
Comment (theengineear) [inline packages/core/src/components/notebook-menu/index.js:414]: 👍

File: packages/core/src/epics/contents.js [hunk 4]
```diff
@@ -79,69 +79,97 @@ export function fetchContentEpic(
...
+        }
+
+        const serverConfig = selectors.serverConfig(host);
+
+        const model = {
+          content: notebook,
+          type: "notebook"
+        };
+
+        // if (action.type === actionTypes.SAVE)
```
Comment (theengineear) [inline packages/core/src/epics/contents.js]: :zap:?

**Review by theengineear — APPROVED** [review]
Cool. I feel like it's a little odd that download / save share the epic as far as organization goes. We could potentially factor out the common code and create two epics... Not sure how you feel about it.

File: packages/core/src/actions.js (hunk 1 above)
Comment (rgbkrk) [reply packages/core/src/actions.js:60]: Yeah I was going a little crazy with the import type then use pattern happening in here. Way simpler to get going and enforce a type on the action creators now. 😄

File: packages/core/src/epics/contents.js (hunk 4 above)
Comment (rgbkrk) [reply packages/core/src/epics/contents.js]: Perhaps I should have left this as a comment instead:

```
// NOTE: assumption is action.type === actionTypes.SAVE at this point
```

**Comment by rgbkrk** [comment]
> Cool. I feel like it's a little odd that download / save share the epic as far as organization goes. We could potentially factor out the common code and create two epics... Not sure how you feel about it.

The amount of boilerplate we've been writing for this stuff has been driving me _slightly_ crazy. I'm going to push up a commit and see what you think.

**Pull request merged**